Skeleton for well-formedness checking of goto models [blocks: #3118, #3147, #3188, #3191, #3187, #3193]

danpoe opened the pull request:
This PR introduces a framework for well-formedness checking of goto models. It also introduces two new macros `DATA_CHECK()` and `DATA_CHECK_WITH_DIAGNOSTICS()` for use in the various validation methods (like `validate()`, `validate_full()` on expressions). Based on the value of a variable `vm` (the validation mode, which is an enum with values `INVARIANT` or `EXCEPTION`), when a well-formedness check fails either a `DATA_INVARIANT()` fails or an `incorrect_goto_program_exceptiont` is thrown.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] My contribution is formatted in line with CODING_STANDARD.md.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

danpoe committed:
Add and use validation methods in equal_exprt

danpoe committed:
Add and use validation methods in code_assignt

danpoe committed:
Add and use validation methods in unsignedbv_typet

danpoe committed:
Add macros for goto model validation

danpoe committed:
Add functions to validate expressions

danpoe committed:
Add functions to validate types

danpoe committed:
Add functions to validate statements

danpoe committed:
Add validation methods to exprt

danpoe committed:
Add validation methods to typet

danpoe committed:
Add validation methods to codet

danpoe committed:
Add validate() methods to (components of) goto models

danpoe committed:
Validate goto models in cbmc (if enabled via command line flag)

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: bcca1a6b151ff49741fda42618de6fd823506f7f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87221686
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

danpoe commented on src/util/validate_expressions.cpp:
```diff
@@ -0,0 +1,79 @@
+  C<expr_type>()(expr, std::forward<Args>(args)...)
+
+template <template <typename> class C, typename... Args>
+void call_on_expr(const exprt &expr, Args &&... args)
```
This function simulates virtual call resolution on `exprt` via a case distinction on `id()`. The structs `call_checkt`, `call_validatet`, and `call_validate_fullt` in `validate.h` specify the method to be called (`check()`, `validate()`, and `validate_full()` resp.) and take as a template parameter the expression type to cast to.

danpoe commented on src/util/validate_expressions.cpp:
```diff
@@ -0,0 +1,79 @@
+  else
+  {
+    CALL_ON_EXPR(exprt);
+  }
```
Cases for more expressions will be added here in the future.

peterschrammel replied:
In the end the last case should fail an invariant in order to actually detect when something is missing. For now it would be good if it could output a warning or so.

danpoe replied:
I've added a macro `REPORT_UNIMPLEMENTED_EXPRESSION_CHECKS` that can be set when compiling. If enabled the ids for currently unimplemented expression checks are reported to `stderr`.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 37a4de43b45402e0b3d44687553b7cb1b57d253a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87236032
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

xbauch commented on src/goto-programs/goto_functions.h:
```diff
@@ -118,6 +118,19 @@ class goto_functionst
+  /// reported via DATA_INVARIANT violations or exceptions.
+  void validate(const namespacet &ns, const validation_modet vm) const
+  {
+    for(const auto &entry : function_map)
```
I have this loop on the `goto_model` level: one less file to modify. But this works just as good.

xbauch commented on src/util/validate.h:
```diff
@@ -0,0 +1,99 @@
+      break;                                                                   \
+    case validation_modet::EXCEPTION:                                          \
+      if(!(condition))                                                         \
+        throw incorrect_goto_program_exceptiont(message);                      \
```
Is there a one-argument constructor for `incorrect_goto_program_exceptiont`?

danpoe committed:
Add check for well-formedness of assign instructions

peterschrammel commented on src/util/validate.h:
```diff
@@ -58,4 +58,42 @@ enum class validation_modet
+template <typename T>
+struct call_checkt
+{
+  static_assert(std::is_base_of<irept, T>::value, "");
```
missing message (several occurrences)

danpoe replied:
Hm, the message wouldn't really add anything here. With C++17 the message will become optional then we could remove it.

peterschrammel commented on src/util/validate_types.cpp:
```diff
@@ -0,0 +1,79 @@
+  }
+  else
+  {
+    CALL_ON_TYPE(typet);
```
see comment for exprt

danpoe replied:
Added a macro `REPORT_UNIMPLEMENTED_TYPE_CHECKS`

peterschrammel commented on src/util/validate_code.cpp:
```diff
@@ -0,0 +1,79 @@
+  }
+  else
+  {
+    CALL_ON_CODE(codet);
```
see comment for exprt

danpoe replied:
added a macro `REPORT_UNIMPLEMENTED_CODE_CHECKS`

peterschrammel commented on src/util/validate_code.cpp:
```diff
@@ -0,0 +1,79 @@
+}
+
+/// Check that the given code statement is well-formed, assuming that all its
+/// enclosed statements, subexpressions, etc. have all ready been checked for
```
already?

peterschrammel commented on src/util/expr.h:
```diff
@@ -210,6 +213,59 @@ class exprt:public irept
+  /// The validation mode indicates whether well-formedness check failures are
+  /// reported via DATA_INVARIANT violations or exceptions.
+  void check(const validation_modet vm = validation_modet::INVARIANT) const
+  {
```
UNREACHABLE? This shouldn't be ever called, I guess? Or remove the method at all.

danpoe replied:
We could add here checks that apply to all `exprt` (but not necessarily to all `irept`). It's currently called by `check_expr_pick()` for cases for which more specific checks have not been implemented yet.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 42bf8fe3b069d26a8d9f375017c18b6da97650ea).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87472451
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

peterschrammel commented on src/util/type.h:
```diff
@@ -74,6 +78,55 @@ class typet:public irept
+  ///
+  /// The validation mode indicates whether well-formedness check failures are
+  /// reported via DATA_INVARIANT violations or exceptions.
+  void check(const validation_modet vm = validation_modet::INVARIANT) const
```
see comment for exprt

danpoe replied:
see reply for exprt

peterschrammel commented on src/util/std_code.h:
```diff
@@ -59,6 +61,49 @@ class codet:public exprt
+  ///
+  /// The validation mode indicates whether well-formedness check failures are
+  /// reported via DATA_INVARIANT violations or exceptions.
+  void check(const validation_modet vm = validation_modet::INVARIANT) const
```
see comment for exprt

danpoe replied:
see reply for exprt :)

peterschrammel commented on src/util/symbol_table.h:
```diff
@@ -109,6 +109,10 @@ class symbol_tablet : public symbol_table_baset
   }
+
+  void validate() const
+  {
```
UNIMPLEMENTED?

peterschrammel replied:
or remove

chrisr-diffblue replied:
I'm working on a PR that will add a body, so I'd leave it empty.

peterschrammel commented on src/util/symbol_table.h:
```diff
@@ -109,6 +109,10 @@ class symbol_tablet : public symbol_table_baset
     return iteratort(internal_symbols.end());
   }
+
+  void validate() const
```
Doxygen

peterschrammel commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -972,6 +978,7 @@ void cbmc_parse_optionst::help()
     " --xml-ui                     use XML-formatted output\n"
     " --xml-interface              bi-directional XML interface\n"
     " --json-ui                    use JSON-formatted output\n"
+    " --validate-goto-model        enables additional well-formedness checks on the goto program\n" // NOLINT(*)
```
break long help text and align below `enables`

danpoe replied:
You mean so the output of `--help` will be at most 80 characters wide, or that the code should be at most 80 characters wide, or both?

xbauch replied:
https://github.com/diffblue/cbmc/blob/7905492ffce5c0f11405474d11cc673c8f02b55f/src/goto-instrument/reachability_slicer.h#L44-L47

danpoe replied:
Thanks @xbauch. So that seems to suggest 80 characters in the output but not necessarily in the code. I'm gonna change it to that.

peterschrammel commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -526,6 +526,12 @@ int cbmc_parse_optionst::doit()
   if(set_properties())
     return CPROVER_EXIT_SET_PROPERTIES_FAILED;
 
+  if(cmdline.isset("validate-goto-model"))
```
Would we like to have this in the other driver programs too?

martin-cs replied:
I would have thought goto-instrument was the place for it.

danpoe replied:
Yes, I think we should have the option in all of the tools to validate the goto model.

@martin-cs I think also in e.g. `cbmc` it would make sense to validate the goto program as a safety net sort of thing. Once right after `goto_convert` and perhaps once more before `goto_symex` to check that all the transformation passes like `remove_function_pointers`, etc. did not produce an ill-formed goto model.

martin-cs replied:
@danpoe : I'm very much in favour of validating the program between all transformation passes; that sounds great.  To the point I might suggest making that the last step of each transformation.  I think this will need to be able to check for various "normal forms" as well, i.e. returns and function pointers removed, etc. ( @tautschnig may have opinions on this subject ).  As I first envisaged this, that should be enough to prove that all of the DATA_INVARIANTs hold.

i also like the "with exceptions when we load it and with invariants after" approach.  My only question is about the specific, user accessible command-line flag to just check the validity of the goto-program and nothing else.  That seems like something we definitely want but that it should probably be in goto-instrument as a "checking and modifying the goto-program" bit of functionality, rather than in every tool.

danpoe replied:
Ah yes, that makes sense. @sonodtt is currently working on some additional checks of the goto program that will allow to pass flags that indicate whether function pointers and returns are expected to have been removed.

martin-cs replied:
Cool.

@sonodtt : having looked at this in the past ( https://github.com/diffblue/cbmc/issues/751 is the first bit I have written down ), may I suggest that you make the flags independent for each feature.  For example, one of the last few differences between the ```static_analyst``` and ```ait``` frameworks is that the later requires the abscence of function pointers (naturally or removed) and return removal.  function pointer removal is an (at the moment undocumented) precondition of return removal and so on.  It would be *amazing* if we could actually map out and enforce these requirements on "normal forms".

martin-cs commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -526,6 +526,12 @@ int cbmc_parse_optionst::doit()
+  if(cmdline.isset("validate-goto-model"))
+  {
+    const namespacet ns(goto_model.symbol_table);
+    goto_model.validate(ns, validation_modet::INVARIANT);
```
Pedantry perhaps but if the design allows for this to fail then it should throw and exception.  Invariants should be reserved for things that should only fail if there are bugs in the code.

xbauch replied:
I guess the idea is to call `validate` with exception once the goto model is loaded and then with invariants after program transformations.

martin-cs replied:
That sounds like a reasonable approach in general.  But doesn't it mean that this specific case should be exceptions?

danpoe replied:
When we do the check here we have already called it once before (in `initialize_goto_model`).

martin-cs replied:
Ah; OK.

martin-cs commented on src/util/std_types.h:
```diff
@@ -1260,6 +1260,12 @@ class unsignedbv_typet:public bitvector_typet
+  void check(const validation_modet vm = validation_modet::INVARIANT) const
+  {
+    DATA_CHECK(
+      !get(ID_width).empty(), "unsigned bitvector type must have width");
```
This condition seems to also appear in ```validate_type```.  Am I missing something or could the duplication be removed?

danpoe replied:
Yes, the `validate_type` function should be removed as it's being replaced by the `check()` method.

martin-cs replied:
OK, is that work scheduled?  It would be good if we could avoid the "we'll clean that up later" effect.

danpoe replied:
Yes, that should be part of this PR. When we introduce a `check()` method we'll replace the corresponding `validate_type()` function.

martin-cs replied:
Excellent!  Thanks.

martin-cs reviewed (commented):
This seems like a useful and a worthwhile thing to do.  I'm not sure I quite 'get' the architecture of this but I think I could be persuaded.  I slightly worry that this might lead to duplication of checks but that is better than not having any at all!

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 41cb8163035cb362c12f80e7286370d4c52be810).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87476584
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 0316f5f8a96061a3c0ed597b4e7ba5494ada8a00).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87480682

chrisr-diffblue commented on src/util/validate.h:
```diff
@@ -0,0 +1,99 @@
+
+/// This macro takes a condition which denotes a well-formedness criterion on
+/// goto programs, expressions, instructions, etc. Based on the value of the
+/// variable vm (the validation mode), it either uses DATA_INVARIANT() to check
```
I may have missed a global declaration of `vm` somewhere, but from what I've seen it looks like `vm` is just 'whatever variable of that name is in scope at the moment?` - Could you either make that explicit in the docs and/or actually add a global declaration for it somewhere? I'm otherwise wondering what the compile error messages might look like if someone tries to use these macros outside of a validate() function, or if the try to declare a variable of a different type but the same name... :-)

danpoe replied:
Ok, I've added some doc and added a static assertion to check that `vm` has the expected type.

chrisr-diffblue commented on src/util/validate.h:
```diff
@@ -0,0 +1,99 @@
+    }                                                                          \
+  } while(0)
+
+#define DATA_CHECK_WITH_DIAGNOSTICS(condition, message, ...)                   \
```
Given that in the exception case only the first member of the varags is supported (and even then, only if its a `source_locationt`), is it better to have a fixed number of args for this macro, rather than varargs? Alternatively, keep it varargs but add a varargs constructor for `incorrect_goto_program_exceptiont` ?

danpoe replied:
I've added a vararg constructor to `incorrect_goto_program_exceptiont`.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 69c46d5be18b3c51612890539dac0e1eb50ad207).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87519725

danpoe committed:
Add new constructor to pass diagnostic information to incorrect_goto_program_exceptiont

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 5befa49ba0f528c347458e85d725edd1007199b3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87533768
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: fd7f4ae31e55510087ae704984035a030bfd02d0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87537826
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: b68b28b1dde001c3b76757f84416a73a0138fc72).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87537737

danpoe committed:
Validate goto models in goto-analyzer (if enabled via command line flag)

danpoe committed:
Validate goto models in goto-instrument (if enabled via command line flags)

There are two flags: --validate-goto-binary reads in a goto binary and checks
that it is well formed and then exits; --validate-goto-model enables additional
well-formedness checks at various points during the execution of goto-instrument
in addition to the normal analyses that it runs.

danpoe commented:
@peterschrammel I think I've addressed all the comments.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 526ec65116d24840a5a82a824aae0c992a6b33e9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87641998
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 29b76b4268704ed9f87f5fac583b4c484edb83dc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87650391

kroening reviewed (changes requested):
The key problem I have with this PR is that it reverses the order of checking and usage: it first does the static cast, and then calls a 'check' method on the casted type. At this point, memory safety is lost.
I would thus advocate that we stick to the order 1) 'check that the preconditions are met' and only then 2) do the cast.

danpoe commented:
@kroening The reasoning was that the precondition for the downcast itself is only that the `id()` field has a certain value. Then we do the downcast and call `check()`.

I think it should be fine for a class to know about and check its own well-formedness conditions. When implementing the checks one should not assume though that some of the conditions already hold. Thus, when implementing the check of an `index_exprt` one should not use e.g. `index()` before having checked that there are two operands. But that's fine because it's within the control of the class itself.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 1150d58540448562ca25409141651a0e4487fe78).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87943340

thk123 commented on src/util/std_expr.h:
```diff
@@ -1414,6 +1415,26 @@ class equal_exprt:public binary_relation_exprt
+
+  void validate(
+    const namespacet &ns,
+    const validation_modet vm = validation_modet::INVARIANT) const
```
:question: `vm` is not used?

danpoe replied:
It's used by the `DATA_CHECK()` macros.

thk123 commented on src/util/std_expr.h:
```diff
@@ -1414,6 +1415,26 @@ class equal_exprt:public binary_relation_exprt
+    check(vm);
+
+    // check types
+    DATA_CHECK(
```
:question: Is there a distinction between "check"ing something and "validate"-ing something? Why are type equality done in validate for example?

danpoe replied:
The distinction between them is documented in `exprt`, `typet`, and `codet`. `check()` is for cheap sanity checks that can be performed frequently, e.g., as part of the `to_<expression-type>` casting functions. `validate()` and `validate_full()` (which descends recursively into the subexpressions and calls `validate()` on them) are for more thorough checks.

sonodtt commented:
Dependency on namespacet is unmet. (in validate.h)

chrisr-diffblue commented:
@kroening I know this PR now has conflicts because develop has moved on, but before we invest any further effort into rebasing this PR, could you please comment on the approach taken here in the light of @danpoe clarifying the approach?

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 582559f067e5f215971ac17fd30163300eac2ee7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90547017

danpoe commented:
The PR has now been adapted according to the suggestions of @kroening in PR #3147.

kroening commented on src/util/std_expr.h:
```diff
@@ -1414,6 +1415,29 @@ class equal_exprt:public binary_relation_exprt
+  {
+    DATA_CHECK(expr.operands().size() == 2, "equality must have two operands");
+  }
+
```
The validation_modet vm doesn't appear to be used? Can it just simply be removed?

thk123 replied:
And were it to be used, it'd probably be a lot easier to have all the checks call `report_validation_error` that can than be configured

danpoe replied:
The validation mode variable `vm` is implicitely used by the `DATA_CHECK()` macros. There's a static assertion in there that checks that it's there and has the correct type.

The `DATA_CHECK()` macros need to be macros as we want the macros like `__LINE__` and `__func__` (which are used by `INVARIANT()`, etc.) to report the location of where the `DATA_CHECK()` macro is in the code.

tautschnig replied:
What is the reason for this _implicit_ use? Why can't we just pass it as another parameter?

danpoe replied:
I just wanted to avoid having to pass the argument every time `DATA_CHECK` is used. But I can make it into a parameter if people prefer it.

tautschnig replied:
I think clarity should win over saving four characters (`vm, `) each time we use it.

kroening commented:
This is now getting there.

One idea, which could make things mildly more beautify (but is functionally equivalent):
a) Consider equal_exprt as an example.
b) The fact that these have two operands stems from the fact that equal_exprt is a binary_exprt.
c) Thus, you could do the check that there are two operands in binary_exprt::check(), and then equal_exprt::check() simply calls binary_exprt::check.

However, if there's urgency, this could be a follow-up PR.

kroening commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -977,6 +983,9 @@ void cbmc_parse_optionst::help()
+    // NOLINTNEXTLINE(whitespace/line_length)
+    " --validate-goto-model        enable additional well-formedness checks on the\n"
+    "                              goto program\n"
     HELP_GOTO_TRACE
```
I'd not add this flag to the help. It's for CBMC developers only, and is a distraction for users.

kroening commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -875,6 +881,9 @@ void goto_analyzer_parse_optionst::help()
+    // NOLINTNEXTLINE(whitespace/line_length)
+    " --validate-goto-model        enable additional well-formedness checks on the\n"
+    "                              goto program\n"
     " --version                    show version and exit\n"
```
(same here: this is noise to users)

thk123 commented:
@kroening 

> I'd not add this flag to the help. It's for CBMC developers only, and is a distraction for users.

:-1: - help is for all users of the program, including developers. Undocumented flags seem like a bad habit to get into! If not relevant to most users perhaps put in a category of "advanced options"?

thk123 commented on src/util/validate.h:
```diff
@@ -0,0 +1,62 @@
+  do                                                                           \
+  {                                                                            \
+    static_assert(                                                             \
+      std::is_same<decltype(vm), const validation_modet>::value,               \
```
:question: Why do this as a macro?

danpoe replied:
The `DATA_CHECK()` macros need to be macros as we want the macros like `__LINE__` and `__func__` (which are used by `INVARIANT()`, etc.) to report the location of where the `DATA_CHECK()` macro is in the code.

thk123 commented on src/util/validate_expressions.cpp:
```diff
@@ -0,0 +1,89 @@
+template <template <typename, typename> class C, typename... Args>
+void call_on_expr(const exprt &expr, Args &&... args)
+{
+  if(expr.id() == ID_equal)
```
:pick: I would use `can_cast_exprt` here to protect you against changes in `id` etc

danpoe replied:
Hm, using `can_cast_exprt` seems to clash a bit with @kroening's view that we should only consider an expression castable to a subtype if all the invariants of the subtype are satisfied (though of course in practice we usually won't check all of the invariants before doing the cast).

thk123 commented on src/util/validate_code.cpp:
```diff
@@ -0,0 +1,90 @@
+#include "validate.h"
+#include "validate_helpers.h"
+
+#define CALL_ON_CODE(code_type)                                                \
```
:question: Why add a macro, what's wrong with calling `call_on_code` directly? (same for expr/type)

danpoe replied:
`call_on_code()` in turn invokes the `CALL_ON_CODE()` macro. I wanted to reduce the noise a bit in the function by introducing the macro.

thk123 reviewed (commented):
I'm very excited for this feature (and apologies for reviewing late in the day - none of my suggestions are blocking) but I hope you agree this needs some tests to just check the errors get threaded through correctly etc. An regression test for the flag (I assume it's possible to construct some malformed goto_model - perhaps by running without the checker, saving a goto binary and feeding that in?) and unit tests for each check would be ideal I think?

tautschnig commented:
> @kroening 
> > I'd not add this flag to the help. It's for CBMC developers only, and is a distraction for users.

@thk123 
> :-1: - help is for all users of the program, including developers. Undocumented flags seem like a bad habit to get into! If not relevant to most users perhaps put in a category of "advanced options"?

From my experience, `--validate-goto-model` would be one of the flags badly needed by users - the amount of trouble that my colleagues run into where we first need to rule out those inconsistencies is why this work is happening in the first place. Of course, in an ideal world we wouldn't need this at all, but as a matter of fact we simply have too many bugs affecting our users.

tautschnig commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -875,6 +881,9 @@ void goto_analyzer_parse_optionst::help()
     "\n"
     "Other options:\n"
+    // NOLINTNEXTLINE(whitespace/line_length)
+    " --validate-goto-model        enable additional well-formedness checks on the\n"
```
Can we have a macro for this, just like `HELP_GOTO_CHECK`?

tautschnig commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -125,8 +125,29 @@ int goto_instrument_parse_optionst::doit()
     get_goto_program();
 
+    {
+      const bool b = cmdline.isset("validate-goto-binary");
```
How about naming this `validate_only` or the likes instead of `b`?

tautschnig commented on src/goto-instrument/goto_instrument_parse_options.h:
```diff
@@ -101,6 +101,8 @@ Author: Daniel Kroening, kroening@kroening.com
   "(show-local-safe-pointers)(show-safe-dereferences)" \
   OPT_REPLACE_CALLS \
+  "(validate-goto-binary)" \
+  "(validate-goto-model)" \
```
Those should also be replaced by macros.

tautschnig commented on src/goto-programs/goto_function.h:
```diff
@@ -110,6 +110,16 @@ class goto_functiont
+  void validate(const namespacet &ns, const validation_modet vm) const
+  {
+    body.validate(ns, vm);
+    validate_type_full_pick(type, ns, vm);
```
I have no idea what "full_pick" might mean?!

danpoe replied:
The function takes a `typet` and then "picks" the concrete subtype such as `unsignedbv_typet` (by looking at `id()`) and then calls `unsignedbv_typet::validate_full(type, ns, vm)`. This is documented in `validate_types.cpp`. Now that you mention it it should probably better be called `validate_full_type_pick()` though.

tautschnig replied:
As a caller, do I actually need to know that it does this "pick" step? Don't I just want to to validate the type, whichever way this is implemented?

danpoe replied:
Good point. I think I'll rename them to `check_type()`, `validate_type()`, and `validate_full_type()` then.

tautschnig commented on src/goto-programs/goto_model.h:
```diff
@@ -89,6 +89,16 @@ class goto_modelt : public abstract_goto_modelt
+  ///
+  /// The validation mode indicates whether well-formedness check failures are
+  /// reported via DATA_INVARIANT violations or exceptions.
+  void validate(const namespacet &ns, const validation_modet vm) const
```
I don't think this should take a `namespacet` as an argument, but instead should always construct it from its own `symbol_table`.

tautschnig commented on src/goto-programs/goto_program.cpp:
```diff
@@ -668,6 +669,26 @@ bool goto_programt::instructiont::equals(const instructiont &other) const
+      "assign instruction should contain an assign statement");
+    DATA_CHECK(targets.empty(), "assign instruction should not have a target");
+    break;
+  default:
```
Please explicitly enumerate all other cases here to make clear what's left to be done, and also avoid that in future some new instruction type is added without a case being inserted here.

danpoe commented:
Btw, @martin-cs has suggested that these checks should be always on, though I think there should be at least the option to turn them off. How do people feel about making the checks always on and providing an option `--no-goto-model-validation` instead to disable them?

As far as documenting that option in the help output goes, I would also prefer if all options (even internal ones) were documented. How about adding a new flag `--help-full` which when enabled first prints the `--help` output and then appends another section `Internal options` to it were options like `--no-simplify`, etc. are documented? (and we may leave off `--help-full` in the output of `--help`).

danpoe committed:
Add macros for common command line interface for goto model validation

allredj reviewed (commented):
:warning:
Diffblue compatibility check is currently unavailable.
Please create manual bump.
(cbmc commit: f51b1f543b516785aba1b801effcbc1edf06e215).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90752927

allredj reviewed (commented):
:warning:
Diffblue compatibility check is currently unavailable.
Please create manual bump.
(cbmc commit: 019f7a2c76374dc85cc426c7c67963b142b75bab).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90754356

tautschnig commented:
May I redirect discussions about `--help` to #547?

peterschrammel reviewed (approved):
Fine with me to get started.

tautschnig commented:
@danpoe Can you please look at the linter and clang-format failures? Otherwise this one is over to @kroening for final review and thus unblocking all the other validation work.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d638ea758677a8df6cbc86e79e3e06ae6860ca87).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90857933

kroening commented on src/util/std_code.h:
```diff
@@ -236,6 +287,39 @@ class code_assignt:public codet
+    DATA_CHECK(
+      base_type_eq(code.op0().type(), code.op1().type(), ns),
+      "lhs and rhs of assignment must have same type");
+  }
```
It is no longer necessary to use base_type_eq to compare types -- you can simply use == !

tautschnig replied:
That's not quite the case just yet, is it? Otherwise we should have a major campaign removing all cases of `base_type_eq`.

kroening replied:
Yes, we should have that campaign, together with getting rid of most instances of ns.follow().

kroening commented on src/util/std_expr.h:
```diff
@@ -849,6 +885,26 @@ class binary_relation_exprt:public binary_predicate_exprt
+    DATA_CHECK(
+      base_type_eq(expr.op0().type(), expr.op1().type(), ns),
+      "lhs and rhs of binary relation expression should have same type");
+  }
```
Same here

kroening commented:
If you use == instead of base_type_eq, then the namespace that gets passed around becomes unused. If that bugs you, you could use it to check whether the tag types are in the symbol table, and whether symbol_exprt identifiers are in the symbol table.

danpoe commented:
Checking that all symbols/tags in the goto program exist in the symbol table and are consistent with the corresponding entries is the topic of @xbauch's PR #3118.

kroening commented:
Ok, if that is meant to remain separate, then you could get rid of the namespacet parameters.
